engine/extended_provider/extended_providers_publisher.go

```diff
@@ -0,0 +1,155 @@
+	"errors"
+	"fmt"
+
+	provider "github.com/filecoin-project/index-provider"
+	"github.com/filecoin-project/storetheindex/api/v0/ingest/schema"
+	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-ipld-prime"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/peer"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// ExtendedProvidersPublisher contains fields and interfaces required for publishing a new ad with Extended Providers
+type ExtendedProvidersPublisher struct {
+	// providerID contains a peer ID of the main provider (the one from the body of the ad)
+	providerID string
+	// privKey contains a private key of the main provider
+	privKey crypto.PrivKey
+	// addrs contains addresses of the main provider
+	addrs []string
+	// engine is an underlying engine that is used for publishing ads
+	engine provider.Interface
+	// providers contains providers' identities, keys, metadata and addresses that can be used as extended providers in the ads.
+	// Those need to be provided only upon initialisation. Subsequent calls to the Pubish API require peer ids only.
+	providers map[string]ExtendedProviderInfo
+}
+
+// ExtendedProviderInfo contains information about extended provider.
+type ExtendedProviderInfo struct {
+	// ID contains peer ID of the extended provider
+	ID string
+	// Metadata contains optional metadata of the extended provider
+	Metadata []byte
+	// Addrs contains a list of extended provider's addresses
+	Addrs []string
+	// Priv contains a provtae key of the extended provider
+	Priv crypto.PrivKey
+}
+
+// NewExtendedProvidersPublisher creates a new ExtendedProvidersPublisher
+func NewExtendedProvidersPublisher(providerID peer.ID, privKey crypto.PrivKey, addrs []ma.Multiaddr, engine provider.Interface, providers ...ExtendedProviderInfo) *ExtendedProvidersPublisher {
+	pub := &ExtendedProvidersPublisher{
+		providerID: providerID.String(),
+		privKey:    privKey,
+		engine:     engine,
+		providers:  map[string]ExtendedProviderInfo{},
+		addrs:      multiaddrsToStrings(addrs),
+	}
+	pub.RegisterOrUpdate(providers...)
+	return pub
+}
+
+// RegisterOrUpdate allows regestering or updating extended provider infos
+func (pub *ExtendedProvidersPublisher) RegisterOrUpdate(providers ...ExtendedProviderInfo) {
+	for _, p := range providers {
+		pub.providers[p.ID] = p
+	}
+}
+
+// Publish verifies, signs and publishes a new extended provider ad to the underlying engine. Identity of the main provider will be appended to the
+// extended provider list automatically and should not be passed in explicitly.
+//
+// - contextID and metadata fields are optional and should be used according to the specification
+// - providers list should contain only those providers whos infos have been registered in the publisher
+func (pub *ExtendedProvidersPublisher) Publish(ctx context.Context, contextID []byte, metadata []byte, override bool, providers ...peer.ID) (cid.Cid, error) {
```

Comment:
Wy not take Advertisement straight up? The number of arguments to this function seems to be a tad high.

If the idea is to assist with construction of `Advertisement` then maybe it would be more intuitive to abstract away `AdvertisementBuilder` instead of `Publish`.

Reply:
I thought about the builder. The issue is that assistance is required not just in the building, but in the signing too. Would your concern be alleviated if instead of passing in providers as a multiarg, I use array instead?

It's just +1 argument to what `NotifyPut` takes currently: 

```
NotifyPut(ctx context.Context, provider *peer.AddrInfo, contextID []byte, md metadata.Metadata) (cid.Cid, error)
```

Reply:
I thought about the builder while jogging and liked the idea :) Pushed the change in.